redfish-exporter/listener.go

```diff
@@ -53,75 +53,88 @@ type OriginOfCondition struct {
 	OdataId string `json:"@odata.id"`
 }
 
-func startListener(AppConfig Config, shutdownChan <-chan struct{}) {
-	var listener net.Listener
-	var err error
+type Server struct {
+	listenIP     string
+	listenPort   string
+	ln           net.Listener
+	shutdownChan chan struct{}
+}
+
+func NewServer(listenIP string, listenPort string) *Server {
+	return &Server{
+		listenIP:     listenIP,
+		listenPort:   listenPort,
+		shutdownChan: make(chan struct{}),
+	}
+}
 
+func (s *Server) Start(AppConfig Config) error {
+	var err error
+	var ln net.Listener
 	if AppConfig.SystemInformation.UseSSL {
 		cert, err := tls.LoadX509KeyPair(AppConfig.CertificateDetails.CertFile, AppConfig.CertificateDetails.KeyFile)
 		if err != nil {
-			log.Fatalf("Failed to load certificates: %v", err)
+			log.Println("Failed to load certificates")
+			return err
 		}
 		config := &tls.Config{Certificates: []tls.Certificate{cert}}
-		listener, err = tls.Listen("tcp", fmt.Sprintf("%s:%d", AppConfig.SystemInformation.ListenerIP, AppConfig.SystemInformation.ListenerPort), config)
+		ln, err = tls.Listen("tcp", fmt.Sprintf("%s:%s", s.listenIP, s.listenPort), config)
 	} else {
-		listener, err = net.Listen("tcp", fmt.Sprintf("%s:%d", AppConfig.SystemInformation.ListenerIP, AppConfig.SystemInformation.ListenerPort))
+		ln, err = net.Listen("tcp", fmt.Sprintf("%s:%s", s.listenIP, s.listenPort))
 	}
 
 	if err != nil {
 		log.Fatalf("Failed to bind to port: %v", err)
+		return err
 	}
 
-	log.Printf("Listening on %s:%d via %s",
-		AppConfig.SystemInformation.ListenerIP,
-		AppConfig.SystemInformation.ListenerPort,
+	s.ln = ln // Store the listener in the Server struct
+
+	go s.acceptLoop(AppConfig)
+
+	log.Printf("Listening on %s:%s via %s",
+		s.listenIP,
+		s.listenPort,
 		func() string {
 			if AppConfig.SystemInformation.UseSSL {
 				return "HTTPS"
 			}
 			return "HTTP"
 		}())
 
-	go func() {
-		<-shutdownChan
-		log.Println("Shutting down listener...")
-		listener.Close()
-	}()
+	<-s.shutdownChan
+	log.Println("Shutting down listener...")
+	ln.Close()
+	return nil
 
+}
+
+func (s *Server) acceptLoop(AppConfig Config) {
 	for {
-		conn, err := listener.Accept()
+		conn, err := s.ln.Accept()
 		if err != nil {
-			select {
-			case <-shutdownChan:
-				return
-			default:
-				log.Printf("Failed to accept connection: %v", err)
-				continue
+			// Check if the error is due to the listener being closed
+			if opErr, ok := err.(*net.OpError); ok && opErr.Err.Error() == "use of closed network connection" {
```

Comment:
i think you can do something like 
```
if errors.Is(err, net.ErrClosed) {
    log.Println("Listener closed, stopping accept loop")
    break
}
```

to avoid stringy comparison